design-proposals/storage-migration/storage-migration.md

```diff
@@ -0,0 +1,282 @@
+
+The user could mix `MigrationVolumes` and MigrationStorageClasses`. For example a volumes of the source storage class specified in MigrationStorageClasses could be present also in `MigrationVolumes`. In this case, the MigrationVolumes takes the precedence and the destination volume will be created by KubeVirt with the information specified for the DestinationPVC. This can be a practical way how the user could overwrite the defaults for a storage class just for a single volume. E.g for overwriting the default naming or the ReclaimPolicySourcePvc.
+
+According to the 'ReclaimPolicySourcePvc' in both scenarios, KubeVirt will determine how to handle the source PVCs. For the `Delete` policy, the source PVC will be automatically deleted while for the `Retain` policy, the PVC will be left untouched.
+
+The core idea is that it is KubeVirt responsability to handle fate of the source PVCs as it is the only component able to orchestrate and set the migration as successfull.
+
+## Design details and motivation
+
+### Motivation on the approaches
+
+From libvirt side there a multiple ways to achieve the storage migration. It is possible to simply copy the storage on the new destination and then swap the original disk wit the new one. This approach is the most perfomant as the overhead is caused only by the storage copy.
+
+A different approach is to simultaneously move the storage and live-migrate the guest. When the destination storage is not available on the host where the VM is executing, this solution is required. Costs are greater since there is a storage migration overhead as well as a live VM migration overhead.
+
+As KubeVirt is designed today and due to Kubernetes design principles of volume pod immutability, it isn't possible to hotplug and unplug new volumes while the pod is running. KubeVirt already supports volume hotplug by attaching new volume to a dummy pod and then by bound mounting the storage in `virt-launcher` pod. Sadly, this method fails to address the issue of removing the old PVCs that were expressly mentioned in the virt-launcher pod.
+Declaring all volumes hotpluggable could be a potential remedy, but we prefer to avoid it because it has been shown to be fragile for some storage providers and opaque to Kubernetes.
+
+The only workable and generic solution for this restriction appears to be live VM migration with storage. Live migration involves the creation of a new virt-launcher pod by KubeVirt that has the destination PVCs attached. The source virt-launcher pod and source PVCs are removed after the migration is finished because they are no longer needed.
+
+This migrating to a new storage provider is an exceptional operation, and very resource intensive. If in the future, the k8s community will consider the volume mutability in the pod spec, then the lighter approach could be taken into consideration.
+
```

Comment:
What about migrating storage for a stopped VM?  Seems like a legit use case to me.  Something users may feel more safe doing.  Obviously doesn't work for WFFC storage.  But should we consider it?  If so, there are a lot of interesting questions there.

Reply:
Right now, I have focused on the most challenging use-case. But this could be also covered by this API. However, the implementation would 2 separate code paths. We could leave it as a further improvement.
Potentially, users can already be achieving it with a CSI clone, but having an update of the volumes of the VMs done by KubeVirt will be surely much more convenient.

Reply:
I think that this API should support stopped VMs and this design proposal should discuss how stopped VMs (or non existing VMIs) are handled.

Reply:
Ok, I'll add this too!